### PR TITLE
JAMES-3830 Jmap Quota/query

### PR DESCRIPTION
**Pull request by vttranlina**
jira: https://issues.apache.org/jira/browse/JAMES-3830
resolve https://github.com/linagora/james-project/issues/4626

### CODE REVIEW

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/QuotaQueryMethod.scala [hunk 1]
```diff
@@ -0,0 +1,64 @@
+package org.apache.james.jmap.method
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/QuotaQueryMethod.scala:1]: license

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/QuotaQueryMethodContract.scala [hunk 2]
```diff
@@ -0,0 +1,852 @@
...
+         |  "methodResponses": [[
+         |    "error",
+         |    {
+         |      "type": "unknownMethod",
+         |      "description":"Missing capability(ies): urn:ietf:params:jmap:quota, urn:ietf:params:jmap:core"
+         |    },
+         |    "c1"]]
+         |}""".stripMargin)
+  }
+}
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/QuotaQueryMethodContract.scala:852]: Test with shared mailboxes?
Test making sure the query does not return somebody else quota?

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/QuotaQueryMethodContract.scala [hunk 3]
```diff
@@ -0,0 +1,852 @@
...
+      .addUser(ANDRE.asString, ANDRE_PASSWORD)
+
+    requestSpecification = baseRequestSpecBuilder(server)
+      .setAuth(authScheme(UserCredential(BOB, BOB_PASSWORD)))
+      .addHeader(ACCEPT.toString, ACCEPT_RFC8621_VERSION_HEADER)
+      .build
+  }
+
+  @Test
+  def queryShouldSuccessByDefault(): Unit = {
```
Comment (quantranhong1999) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/QuotaQueryMethodContract.scala:64]:
```suggestion
  def queryShouldSucceedByDefault(): Unit = {
```

**Review by Arsnael — APPROVED** [review]

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/QuotaQueryMethodContract.scala [hunk 4]
```diff
@@ -849,4 +850,120 @@ trait QuotaQueryMethodContract {
...
+      .body(
+        s"""{
+           |  "using": [
+           |    "urn:ietf:params:jmap:core",
+           |    "urn:ietf:params:jmap:quota",
+           |    "urn:apache:james:params:jmap:mail:shares" ],
+           |  "methodCalls": [[
+           |    "Quota/query",
+           |    {
+           |      "accountId": "29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6",
```
Comment (quantranhong1999) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/QuotaQueryMethodContract.scala:874]: Making this `29883...` constant say `BOB_ACOUNT_ID` will help the tests more understandable IMO.
Currently, I don't know if it is Bob or Andre who is querying...

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/json/QuotaSerializer.scala [hunk 5]
```diff
@@ -42,6 +43,23 @@ object QuotaSerializer {
   private implicit val stateWrites: Writes[UuidState] = Json.valueWrites[UuidState]
 
   private implicit val resourceTypeWrite: Writes[ResourceType] = resourceType => JsString(resourceType.asString())
+  private implicit val scopeReads: Reads[Scope] = {
+    case JsString("account") => json.JsSuccess(AccountScope)
+    case _ => JsError(s"Expecting a JsString to represent a scope property")
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/json/QuotaSerializer.scala:48]: We caan refine error handling here.

If I pass a wrong JsString it returns me "please send me a JsString" while it should say "Oops, I do not manage this value"

```suggestion
    case JsString("account") => json.JsSuccess(AccountScope)
    case JsTring(any) => JsError(s"Unexpected value $any, only 'account' is managed")
    case _ => JsError(s"Expecting a JsString to represent a scope property")
```

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/json/QuotaSerializer.scala [hunk 6]
```diff
@@ -42,6 +43,23 @@ object QuotaSerializer {
...
+  }
+
+  private implicit val optionReads: Reads[QuotaName] = Json.valueReads[QuotaName]
+
+  private implicit val resourceTypeReads: Reads[ResourceType] = {
+    case JsString("count") => json.JsSuccess(CountResourceType)
+    case JsString("octets") => json.JsSuccess(OctetsResourceType)
+    case _ => JsError(s"Expecting a JsString to represent a resourceType property")
+  }
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/json/QuotaSerializer.scala:57]: Idem 

```suggestion
  private implicit val resourceTypeReads: Reads[ResourceType] = {
    case JsString("count") => json.JsSuccess(CountResourceType)
    case JsString("octets") => json.JsSuccess(OctetsResourceType)
    case JsTring(any) => JsError(s"Unexpected value $any, only 'count' and 'octets' are managed")
    case _ => JsError(s"Expecting a JsString to represent a resourceType property")
  }
```

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/json/QuotaSerializer.scala [hunk 7]
```diff
@@ -42,6 +43,23 @@ object QuotaSerializer {
...
+  private implicit val dataTypeReads: Reads[DataType] = {
+    case JsString("Mail") => json.JsSuccess(MailDataType)
+    case _ => JsError(s"Expecting a JsString to represent a dataType property")
+  }
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/json/QuotaSerializer.scala:61]: Idem


```suggestion
  private implicit val dataTypeReads: Reads[DataType] = {
    case JsString("Mail") => json.JsSuccess(MailDataType)
     case JsTring(any) => JsError(s"Unexpected value $any, only 'Mail' are managed")
    case _ => JsError(s"Expecting a JsString to represent a dataType property")
  }
```

**Review by chibenwa — COMMENTED** [review]
Please document that FilterConditions AND and OR are not supported for Quota/query ...  CF https://jmap.io/spec-core.html#query

Idem for position, limit & all complex arguments of /query...

File: server/protocols/jmap-rfc-8621/doc/specs/spec/quotas/quota.mdown [hunk 8]
```diff
@@ -79,10 +79,9 @@ A **FilterCondition** object has the following properties, any of which may be o
 A Quota object matches the FilterCondition if and only if all of the given conditions match. If zero properties are
 specified, it is automatically true for all objects.
 
-The following Quota properties MUST be supported for sorting:
+The FilterOperator (AND/OR/NOT) is not supported
 
-* **name**
-* **used**
+The following properties are not supported: sort, position, anchor, anchorOffset, limit, calculateTotal.
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621/doc/specs/spec/quotas/quota.mdown:84]: You modified the spec here! That is bad!

```
The following Quota properties MUST be supported for sorting:

* **name**
* **used**

> :warning:
> Not implemented
> The following properties are not supported: sort, position, anchor, anchorOffset, limit, calculateTotal.
```

BTW we should likely support the `sort` option with the two mentionned properties...

Comment (vttranlina) [reply server/protocols/jmap-rfc-8621/doc/specs/spec/quotas/quota.mdown:84]: > BTW we should likely support the sort option with the two mentionned properties...

We don't support `limit, offset`, so `sort` don't make sense.
The list returns only `ids` value in the array json

Comment (chibenwa) [reply server/protocols/jmap-rfc-8621/doc/specs/spec/quotas/quota.mdown:84]: Ok but do not modify the spec, annotate it!

File: server/protocols/jmap-rfc-8621/doc/specs/spec/quotas/quota.mdown [hunk 9]
```diff
@@ -79,9 +79,14 @@ A **FilterCondition** object has the following properties, any of which may be o
...
-The FilterOperator (AND/OR/NOT) is not supported
+The following Quota properties MUST be supported for sorting:
 
-The following properties are not supported: sort, position, anchor, anchorOffset, limit, calculateTotal.
+* **name**
+* **used**
+
+> :warning:
+> Not implemented
+> The following properties are not supported: sort, position, anchor, anchorOffset, limit, calculateTotal.
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621/doc/specs/spec/quotas/quota.mdown:89]:
```suggestion
> The following properties are not supported: sort, position, anchor, anchorOffset, limit, calculateTotal.
> FilterOperators (AND/OR/NOT) are not supported.
```

**Review by chibenwa — APPROVED** [review]

**Review by Arsnael — APPROVED** [review]

**Review by quantranhong1999 — APPROVED** [review]

**Comment by vttranlina** [comment]
squash fixup

**Pull request merged**